cobaya/tools.py

```diff
@@ -1069,3 +1069,251 @@ def sort_cosmetic(info):
+    ``N``-tuple ``x``.
+
+    Tuples are sorted internally, and then by ascending order of their values.
+
+    Tuples are uniquified internally up to machine precision, and an adaptive
+    tolerance (relative to min absolute and relative differences in the list) is
+    applied at retrieving.
+
+    Adaptive tolerance is defined between limits ``[atol|rtol]_[min|max]``.
+    """
+
+    values: np.ndarray
+
+    def __init__(self, values=[],
+                 rtol_min=1e-5, rtol_max=1e-3, atol_min=1e-8, atol_max=1e-6, logger=None):
+        assert rtol_min <= rtol_max, \
+            f"{rtol_min=} must be smaller or equal to {rtol_max=}"
+        assert atol_min <= atol_max, \
+            f"{atol_min=} must be smaller or equal to {atol_max=}"
+        self.atol_min, self.atol_max = atol_min, atol_max
+        self.rtol_min, self.rtol_max = rtol_min, rtol_max
+        if logger is None:
+            self.log = get_logger(self.__class__.__name__)
+        else:
+            self.log = logger
+        self.update(values)
+
+    @property
+    def d(self):
+        return len(self.values.shape)
+
+    def __len__(self):
+        return len(self.values)
+
+    def _update_tolerances(self):
+        """Adapts tolerance to the differences in the list."""
+        if self.d == 1:
+            # Assumes that the pool is sorted!
+            values = self.values
+        else:
+            values = np.copy(self.values)
+            values.flatten()
+            values = combine_1d(values)
+        if len(values) > 1:
+            differences = values[1:] - values[:-1]
+            min_difference = np.min(differences)
+            self._adapt_atol_min = self.atol_min * min_difference
+            self._adapt_atol_max = self.atol_max * min_difference
+            min_rel_difference = np.min(differences / values[1:])
+            self._adapt_rtol_min = self.rtol_min * min_rel_difference
+            self._adapt_rtol_max = self.rtol_max * min_rel_difference
+        else:  # single-element list
+            self._adapt_atol_min = self.atol_min * values[0]
+            self._adapt_atol_max = self.atol_max * values[0]
+            self._adapt_rtol_min = self.rtol_min
+            self._adapt_rtol_max = self.rtol_max
+
+    def update(self, values):
+        """Adds a set of values, uniquifies and sorts."""
+        values = self._check_values(values)
+        self._update_values(values)
+        self._update_tolerances()
+
+    @abstract
+    def _check_values(self, values):
+        """
+        Checks that the input values are correctly formatted and re-formats them if
+        necessary.
+
+        Returns a correctly formatted array.
+
+        Internal sorting is enforced, but external is is ignored.
+        """
+
+    @abstract
+    def _update_values(self, values):
+        """Combines given and existing pool. Should assign ``self.values``."""
+
+    def find_indices(self, values):
```

Comment:
Is this much slower than the old method for large arrays? [seems to be doing an in-python loop]

Reply:
The old method would return wrong values for input values with a small positive epsilon added, so not a fair comparison. In any case, it depends on slow is much slower. With this test:

    from cobaya.tools import Pool1D
    import numpy as np
    n = 500
    n_test = 100
    values = np.random.random(n)
    pool = Pool1D(values)
    test_values = np.random.choice(values, n_test) + 1e-15 * np.random.random(n_test)
    %timeit pool.find_indices(test_values)

One can check that the times are:

    find 100 in 500: ~10ms
    find 1000 in 5000: ~50ms
    find 10000 in 50000: ~1.3s

which looks reasonable up to the few thousands, I'd say, given the expected costs of the rest of the pipeline. For the >10k case, we could even make it faster (see note in the `PoolND` code about taking advantage of the pool being sorted). If needed, we can even rewrite the relevant part in cython/numba.

Reply:
I'm sure a few thousand is enough. But e.g. could compare with doing old method first and then only doing the new method if the exact value is not found (or more generally, test a hash index first for an exact match, and only search if not found)

Reply:
Good idea! https://github.com/CobayaSampler/cobaya/pull/222/commits/38454acc1af4cd338c17321238ed890c83508b4f does the fast search first, and the slow one for the elements not found in the first round. As fast as the original one if elements are actually equal, but slower if small perturbations introduced. Still guaranteed correct (or if not, at least not failing silently!). Now I need to do it for 2D.

Reply:
Done for 2D too (still has a python loop, but only over test values. At least 10x faster than the slow and coarse alternative). Added non-cosmo test for 1D and 2D